src/nn/batchnorm1d.rs

```diff
@@ -167,61 +167,45 @@ impl<B: Dim, const C: usize, E: Dtype, D: Device<E>>
     }
 }
 
-impl<const C: usize, E: Dtype, D: Device<E>> BuildModule<D, E> for BatchNorm1D<C, E, D> {
-    fn try_build(device: &D) -> Result<Self, D::Err> {
-        Ok(Self {
-            scale: device.try_ones()?,
-            bias: device.try_zeros()?,
-            running_mean: device.try_zeros()?,
-            running_var: device.try_ones()?,
-            epsilon: E::from_f32(1e-5).unwrap(),
-            momentum: E::from_f32(0.1).unwrap(),
-        })
-    }
-}
-
 impl<const C: usize, E: Dtype, D: Device<E>> TensorCollection<E, D> for BatchNorm1D<C, E, D> {
-    fn iter_tensors<V: ModuleVisitor<Self, E, D>>(visitor: &mut V) -> Result<(), V::Err> {
-        visitor.visit_tensor(
+    type Output<E2: Dtype, D2: Device<E2>> = BatchNorm1D<C, E2, D2>;
+
+    fn iter_tensors<E2: Dtype, D2: Device<E2>, V: ModuleVisitor<Self, E, D, E2, D2>>(
+        visitor: &mut V,
+    ) -> ModuleVisitorOutput<V::Func, Self, E, D, E2, D2> {
+        let scale = visitor.visit_tensor(
             "scale",
             |s| &s.scale,
             |s| &mut s.scale,
             TensorOptions::reset_to_ones(),
         )?;
-        visitor.visit_tensor(
+        let bias = visitor.visit_tensor(
             "bias",
             |s| &s.bias,
             |s| &mut s.bias,
             TensorOptions::reset_to_zeros(),
         )?;
-        visitor.visit_tensor(
+        let running_mean = visitor.visit_tensor(
             "running_mean",
             |s| &s.running_mean,
             |s| &mut s.running_mean,
             TensorOptions::detached(|t| t.try_fill_with_zeros()),
         )?;
-        visitor.visit_tensor(
+        let running_var = visitor.visit_tensor(
             "running_var",
             |s| &s.running_var,
             |s| &mut s.running_var,
             TensorOptions::detached(|t| t.try_fill_with_ones()),
-        )
-    }
-}
+        )?;
 
-impl<const C: usize, E: Dtype, D1: Device<E>, D2: Device<E>> ToDevice<D2>
-    for BatchNorm1D<C, E, D1>
-{
-    type Output = BatchNorm1D<C, E, D2>;
-    fn to_device(&self, device: &D2) -> Self::Output {
-        BatchNorm1D {
-            scale: self.scale.to_device(device),
-            bias: self.bias.to_device(device),
-            running_mean: self.running_mean.to_device(device),
-            running_var: self.running_var.to_device(device),
-            epsilon: self.epsilon,
-            momentum: self.momentum,
-        }
+        Ok(crate::try_some!(BatchNorm1D {
```

Comment:
is `crate::try_some` just doing this?
```suggestion
        Ok(Some(BatchNorm1D {
```

Reply:
The main purpose of try_some is to wrap that code in a closure that returns an Option, so that the ? operator can be used to propagate Nones. We need to use the ? operator twice on the results so that we short circuit any errors immediately, while not skipping over any tensors when we aren't outputting a Module.